task-impls/src/da.rs

```diff
@@ -286,8 +339,84 @@ where
+                if *self.cur_view >= *view {
+                    return None
+                }
                 self.cur_view = view;
+
+                // TODO ED Make this a new task so it doesn't block main DA task
+
+                // If we are not the next leader (DA leader for this view) immediately exit
+                if self.committee_exchange.get_leader(self.cur_view + 1)
+                    != self.committee_exchange.public_key().clone()
+                {
+                    return None;
+                }
+
+                // ED Copy of parent_leaf() function from sequencing leader
+
+                let parent_view_number = &self.high_qc.view_number;
+                let consensus = self.consensus.read().await;
+                let Some(parent_view) = consensus.state_map.get(parent_view_number) else {
+                    warn!("Couldn't find high QC parent in state map.");
+                    return None;
+                };
+                let Some(leaf) = parent_view.get_leaf_commitment() else {
+                    warn!(
+                        ?parent_view_number,
+                        ?parent_view,
+                        "Parent of high QC points to a view without a proposal"
+                    );
+                    return None;
+                };
+                let Some(leaf) = consensus.saved_leaves.get(&leaf) else {
+                    warn!("Failed to find high QC parent.");
+                    return None;
+                };
+                let parent_leaf = leaf.clone();
+
+                // Prepare the DA Proposal
+                //         let Some(parent_leaf) = self.parent_leaf().await else {
+                //     warn!("Couldn't find high QC parent in state map.");
+                //     return None;
+                // };
+
+                drop(consensus);
+
+                let mut block = <TYPES as NodeType>::StateType::next_block(None);
+                let txns = self.wait_for_transactions(parent_leaf).await?;
+
+                for txn in txns {
+                    if let Ok(new_block) = block.add_transaction_raw(&txn) {
+                        block = new_block;
+                        continue;
+                    }
+                }
+                let block_commitment = block.commit();
+
+                let consensus = self.consensus.read().await;
+                let signature = self.committee_exchange.sign_da_proposal(&block.commit());
+                let data: DAProposal<TYPES> = DAProposal {
+                    deltas: block.clone(),
+                    view_number: self.cur_view,
+                };
+                let message = SequencingMessage::<TYPES, I>(Right(
+                    CommitteeConsensusMessage::DAProposal(Proposal { data, signature }),
+                ));
+                // Brodcast DA proposal
+                // TODO ED We should send an event to do this, but just getting it to work for now
```

Comment:
Isn't the event just the `SendDABlockData` below? Or do we need an additional event?

Reply:
The SendDABlockData event is only for the DA leader in the DA task to send the block data to itself in the consensus task.  The DAProposalSend event is to send the proposal to the DA committee over the network.  There might be a better way to do this in the future, but sending an event seemed like the easiest way to get things working.